Add a simple salt unit tests for `metalk8s_network` custom Salt module

TeddyAndrieux opened the pull request:
**Component**:

'salt', 'tests', 'ci'
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

Add simple Salt unit tests
#1136 

**Summary**:

- Split unit_tests stage to run unit tests simultaneously
- Add unit tests for `metalk8s_network` salt custom module
- Add Salt unit tests in pre-merge

---

Fixes: #1136

TeddyAndrieux committed:
eve: Split `unit_tests` stage of CI

Split unit_tests stage to run unit test from UI and storage-operator
simultaneously

TeddyAndrieux committed:
salt/tests: Add simple unit_tests for `metalk8s_network`

Add a simple unit tests for `metalk8s_network` salt custom module, this
tests use `salttesting` and pytest

Sees: #1136

TeddyAndrieux committed:
tox: Add `unit-tests` step in tox for Salt unit test

Add `unit-tests` to tox so that we can run Salt unit tests easily, we
force env to `python2` as salt currently run with python2 so we need to
run test in python2 (we also need to force python2 for pip-compile of
this specific requirements (so add a `pip-compile2` to tox as well)

Sees: #1136

TeddyAndrieux committed:
eve: Add `unit_tests_salt` in pre-merge in CI

Add a stage in pre-merge to run Salt unit-tests

Sees: #1136

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/GH-1136-add-simple-salt-unit-tests` into
`development/2.4`
* [w/2.5/improvement/GH-1136-add-simple-salt-unit-tests](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/GH-1136-add-simple-salt-unit-tests) will be merged into `development/2.5` 
* [w/2.6/improvement/GH-1136-add-simple-salt-unit-tests](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/GH-1136-add-simple-salt-unit-tests) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

gdemonet commented on eve/main.yml:
```diff
@@ -914,17 +925,48 @@ stages:
+        docker-unit-tests:
+          context: 'storage-operator'
+          dockerfile: eve/workers/pod-unit-tests/storage-operator/Dockerfile
+      vars:
+        name: "metalk8s-unit-tests-storage-operator-worker"
```
Ohhhhh I didn't know we could do this! Quite neat :)

bert-e commented:
# History mismatch



Merge commit #a352520276f2b83b1976db15655325505618f471 on the integration branch
`w/2.5/improvement/GH-1136-add-simple-salt-unit-tests` is merging a branch which is neither the current
branch `improvement/GH-1136-add-simple-salt-unit-tests` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/GH-1136-add-simple-salt-unit-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/GH-1136-add-simple-salt-unit-tests` into
`development/2.4`
* [w/2.5/improvement/GH-1136-add-simple-salt-unit-tests](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/GH-1136-add-simple-salt-unit-tests) will be merged into `development/2.5` 
* [w/2.6/improvement/GH-1136-add-simple-salt-unit-tests](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/GH-1136-add-simple-salt-unit-tests) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

gdemonet commented on eve/workers/pod-unit-tests/ui/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+FROM centos:7
```
Out of scope for this PR, but we may want to open a debt ticket (cc @ChengYanJin) to make sure this worker has run `npm ci` and is only rebuilt when the `package-lock.json` changes. Would reduce the impact of these tests when nothing changed in `ui/`.

gdemonet commented on salt/tests/unit/modules/conftest.py:
```diff
@@ -0,0 +1,11 @@
+# Add our Salt module directory to the python path
+sys.path.insert(
+    0,
+    os.path.join(
+        os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))),
+        "_modules"
+    )
```
Could be easier (IMHO) to read with:

```suggestion
    os.path.realpath(os.path.join(
        __file__, "..", "..", "..", "..", "_modules"
    ))
```

TeddyAndrieux replied:
It will be more something like this (since `__file__` is a file and not a directory)
```suggestion
    os.path.realpath(os.path.join(
        os.path.dirname(__file__), "..", "..", "..", "_modules"
    ))
```
Personally prefer the current approach ...

gdemonet replied:
Fine by me, anyway as soon as we get Python 3 we'll rewrite it with `pathlib` :stuck_out_tongue_winking_eye:

gdemonet commented on salt/tests/unit/modules/test_metalk8s_network.py:
```diff
@@ -0,0 +1,87 @@
+from salttesting.unit import TestCase
+from salttesting.mock import MagicMock, patch
+
+import metalk8s_network
+
+
+class Metalk8sNetworkTestCase(TestCase, LoaderModuleMockMixin):
```
So with this `TestCase` base class, we get `unittest`-style test cases? Not `pytest`-style?

TeddyAndrieux replied:
Yes

gdemonet replied:
OK, so everything related to fixtures is implemented through mixins, and assertions need to use the "right" methods.

I'm OK with this choice, especially if the tooling provided by Saltstack forces us to go this way. But I would like the squad @scality/metalk8s to acknowledge this decision before we move forward, as this is different from the other `pytest` suite we have.

alexandre-allard replied:
Looks pretty clear to me, at least with these simple tests.

gdemonet commented on salt/tests/unit/modules/test_metalk8s_network.py:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+
+    def test_get_kubernetes_service_ip_success(self):
```
There's quite a bit of duplication between tests for both methods. Can we parametrize test cases with the proposed approach?

TeddyAndrieux replied:
With pytest `paramtrize` nop but with `parameterized` yes sure, but here I don't see any benefits it's 2 differents function we test here even if the result are the same parametrized do not look really usefull, we could define some variable for error message but that pretty much it to me

gdemonet commented on eve/main.yml:
```diff
@@ -969,6 +970,41 @@ stages:
+      path: eve/workers/pod-unit-tests/pod.yaml
+      images:
+        docker-unit-tests:
+          context: '.'
```
Stray copy-pasta?

gdemonet reviewed (commented):
Early comments/questions

bert-e commented:
# History mismatch



Merge commit #cc6770863599d84ec007aa2ed84f30daf5436076 on the integration branch
`w/2.5/improvement/GH-1136-add-simple-salt-unit-tests` is merging a branch which is neither the current
branch `improvement/GH-1136-add-simple-salt-unit-tests` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/GH-1136-add-simple-salt-unit-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/GH-1136-add-simple-salt-unit-tests` into
`development/2.4`
* [w/2.5/improvement/GH-1136-add-simple-salt-unit-tests](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/GH-1136-add-simple-salt-unit-tests) will be merged into `development/2.5` 
* [w/2.6/improvement/GH-1136-add-simple-salt-unit-tests](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/GH-1136-add-simple-salt-unit-tests) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

gdemonet reviewed (commented):
One comment around the choice of tooling, which, since different from usual, should have agreement from the team.

And a small request (could be done in a follow-up PR if you prefer, should wait for agreement anyway): can we provide a minimal guide (with references to Salt guides where needed) to write more of those tests?

TeddyAndrieux committed:
salt/tests: Add simple `README.md` for Salt testing

Add a simple README for Salt testing, only containing simple explanation
for unit tests

alexandre-allard commented on salt/tests/README.md:
```diff
@@ -0,0 +1,23 @@
+## Unit
+Unit tests are used to test each Salt custom functions in MetalK8s.
+
+It use [unittest](https://docs.python.org/2/library/unittest.html)
```
```suggestion
It uses [unittest](https://docs.python.org/2/library/unittest.html)
```

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/GH-1136-add-simple-salt-unit-tests` into
`development/2.4`
* [w/2.5/improvement/GH-1136-add-simple-salt-unit-tests](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/GH-1136-add-simple-salt-unit-tests) will be merged into `development/2.5` 
* [w/2.6/improvement/GH-1136-add-simple-salt-unit-tests](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/GH-1136-add-simple-salt-unit-tests) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

Ebaneck commented on salt/tests/requirements.in:
```diff
@@ -0,0 +1,5 @@
+pytest ~= 4.3.0
+salt == 3000.3
+salttesting == 2018.9.21
```
Maybe in future, we could consider adding flavour by using salts own pylint (saltpylint)

TeddyAndrieux replied:
Not really it's not a requirement for tests but for linting and it's part of this issue more #2649

Ebaneck replied:
Ok agreed! so `saltpylint` should also be considered as a potential linting tool

Ebaneck commented on salt/tests/README.md:
```diff
@@ -0,0 +1,23 @@
+# Salt Tests
+
+## Unit
+Unit tests are used to test each Salt custom functions in MetalK8s.
```
```suggestion
Unit tests are used to test each Salt custom function in MetalK8s.
```

Ebaneck commented on salt/tests/README.md:
````diff
@@ -0,0 +1,23 @@
+### Run tests using tox
+
+```
+tox -e unit-tests
````
No room for markers here I think...
So each time we initiate this, we must run the entire test suite?

TeddyAndrieux replied:
I just need to change the tox.ini and you can do classic pytest filtering like `tox -e unit-tests -k "not kubernetes"` (that filter out every test with `kubernetes` in the name)

bert-e commented:
# History mismatch



Merge commit #cc7f452b6978c092d7d354b384ed23fe02e5035c on the integration branch
`w/2.5/improvement/GH-1136-add-simple-salt-unit-tests` is merging a branch which is neither the current
branch `improvement/GH-1136-add-simple-salt-unit-tests` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/GH-1136-add-simple-salt-unit-tests` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/GH-1136-add-simple-salt-unit-tests` into
`development/2.4`
* [w/2.5/improvement/GH-1136-add-simple-salt-unit-tests](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/GH-1136-add-simple-salt-unit-tests) will be merged into `development/2.5` 
* [w/2.6/improvement/GH-1136-add-simple-salt-unit-tests](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/GH-1136-add-simple-salt-unit-tests) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

TeddyAndrieux commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/0bdb0239efbd32a40a5f90361723de5fab85b1b8) did not succeed in branch improvement/GH-1136-add-simple-salt-unit-tests.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/8e982cdf930499b94c6943726dedb85c84c30ceb) did not succeed in branch w/2.5/improvement/GH-1136-add-simple-salt-unit-tests.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue GH-1136.

Goodbye teddyandrieux.
